source/common/ssl/context_config_impl.h

```diff
@@ -54,18 +55,29 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
   unsigned minProtocolVersion() const override { return min_protocol_version_; };
   unsigned maxProtocolVersion() const override { return max_protocol_version_; };
 
+  bool isValid() const override {
+    // either secret_provider_ is nullptr or secret_provider_->secret() is NOT nullptr.
+    return !secret_provider_ || secret_provider_->secret();
```

Comment:
Where does this function actually end up getting called. Is it always on the main thread? Or will it be called from multiple threads? I'm not sure this is thread safe. Can you explain the full flow? This ends up being quite confusing.

Reply:
It is called from the main thread.  It is called in the ContextManagerImpl::createClientContext() and createServerContext(),  if pass-in ContextConfig is not valid,  return nullptr for SocketFactory. 

In the next PR,  SocketFactory will register a UpdateCallback to SDS_api,   if SDS secret is fetched,  it will call the onUpdate(), SocketFactory will call ContextManagerImpl::createClientContext() again to create a valid Context if ContextConfig is valid.  This is called in the main thread too for this update case since SDS_api onConfigUpdate will be called in the main thread.

Reply:
OK thanks. I think we will need proper locking around the shared_ptr swaps and we might want to use TLS at a later time, but this should work for now and we can figure it out in the follow up PR.